15 alert on leave page

mtdamert opened the pull request:
## Overview

When the user tries to leave the Review or Generate page before saving changes, an alert is popped up.

### Does this close any currently open issues?

<!-- Change ### to #[number of issue], e.g. #1 -->
Closes #15

### Testing Instructions

Here are the cases I found to check:
- browser refresh
- browser tab close
- Log Out button clicked
- Home button in top left clicked
- Review Petitions button clicked
- for a multi-page petition:
  * Previous Petition
  * Save & Continue
  * Review Petitions

When the user tries to save or move on, an alert should not pop up.

mtdamert committed:
#15 - alert on leaving generate page

joshuagerstein commented on frontend/src/src/components/ReviewPage/index.js:
```diff
@@ -79,6 +83,12 @@ export default function ReviewPage(props) {
         })
     }, [petitioner, petitions])
 
+    function startNewPetition() {
```
This is good, a similar function should be made for the Generate component's new petition button (which only shows up if there's exactly one petition)

joshuagerstein commented on frontend/src/src/components/util/navBlock.js:
This component works well. React functions starting with "use" are conventionally custom hooks, would you be open to renaming the component to something like NavBlock?

joshuagerstein committed:
refactor: Centralize navigation blocking state with Context

This builds on the initial navigation blocking feature by centralizing its state into a shared React Context, allowing global components like the Nav bar to interact with the block before logging out.

* A new NavBlockContext (navBlockContext.jsx) is introduced to manage and provide a single, shared navigation blocking state (blockNavRef)

* The GeneratePage and ReviewPage components are refactore to use this new shared context instead of their previous local refs

* The Nav component's logout function now uses the context to check the block status and confirm with the user before logging out

* Added NavBlock to /upload

* The PageNotFound route in App.js is changed from a PrivateRoute to a standard Route to make it accessible to unauthenticated users

joshuagerstein committed:
Merge branch 'develop' into 15-alert-on-leave-page

dnlsngr reviewed (approved):
Ok, this looks good to me. It's a bit of an aggressive requirement to not just be guarding in the context of a modal, but rather in an entire app, which explains having to keep track of this in context. But accomplishes what you set out to, so nice job figuring it out
